### PR TITLE
Factor out cache implementation to standalone package.

### PR DESCRIPTION
**Pull request by mikehelmick**
* Create write thru cache implementation
* Move database provider to it
* First part of #555

Follow on PR will bring this to other parts of the codebase that use caching.

### CODE REVIEW

File: internal/cache/cache.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+	return &Cache{
+		data: make(map[string]item),
+	}
+}
+
+func (c *Cache) purgeExpired(name string, exp int64) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if item, ok := c.data[name]; ok && item.expiresAt == exp {
```
Comment (sethvargo) [inline internal/cache/cache.go:53]: This shouldn't be ==, right? item.expired()?

File: internal/cache/cache.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+	val, hit := c.Lookup(name)
+	if hit {
+		return val, nil
+	}
+
+	// Escalate the lock to a RW lock.
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	// double check that the value hasn't been set by another goroutine.
+	if val, hit := c.data[name]; hit && !val.expired() {
```
Comment (sethvargo) [inline internal/cache/cache.go:81]: It might be helpful to separate these apart for readability.

File: internal/cache/cache.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+}
+
+// Lookup checks the cache for a non-expired object by the supplied key name.
+// The bool return informs the caller if there was a cache hit or not.
+// A return of nil, true means that nil is in the cache.
+// Where nil, false indicates a cache miss or that the value is expired and should
+// be refreshed.
+func (c *Cache) Lookup(name string) (interface{}, bool) {
+	c.mu.RLock()
```
Comment (sethvargo) [inline internal/cache/cache.go:107]: If we take out a full lock here, we can delete from the map without spawning the background goroutine.

File: internal/cache/cache.go (hunk 1 above)
Comment (sethvargo) [reply internal/cache/cache.go:53]: Oh I see after reading the code below. The behavior and purpose of this is unclear. I think it should either accept a name and item or be moved inline.

File: internal/cache/cache.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	object    interface{}
+	expiresAt int64
+}
+
+func (i *item) expired() bool {
+	return i.expiresAt < time.Now().UnixNano()
+}
+
+// New creates a new in memory cache.
+func New() *Cache {
```
Comment (sethvargo) [inline internal/cache/cache.go:43]: Expiration feels like it should be a property of the cache, not a property of each individual cache item. WDYT?

File: internal/cache/cache.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+import (
+	"errors"
+	"sync"
+	"time"
+)
+
+var ErrInvalidDuration = errors.New("expireAfter duration cannot be negative")
+
+type WriteThru func() (interface{}, error)
```
Comment (sethvargo) [inline internal/cache/cache.go:26]: Can we just name this Func (`cache.Func`)?

File: internal/cache/cache_test.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	}
+	if err.Error() != "nope" {
+		t.Errorf("incorrect error, want: `nope` got: %v", err.Error())
+	}
+	if got != nil {
+		t.Errorf("unexpecetd cached item, want: nil, got %v", got)
+	}
+}
+
+func TestInvalidDuration(t *testing.T) {
```
Comment (sethvargo) [inline internal/cache/cache_test.go:121]: Can you add a test for concurrency using two goroutines please?

File: internal/cache/cache.go (hunk 4 above)
Comment (mikehelmick) [reply internal/cache/cache.go:43]: for our purposes, duration per instance is sufficient.

File: internal/cache/cache.go (hunk 1 above)
Comment (mikehelmick) [reply internal/cache/cache.go:53]: added additional comments + changed name of exp

File: internal/cache/cache.go (hunk 2 above)
Comment (mikehelmick) [reply internal/cache/cache.go:81]: tried a few ways - i like this the best.

File: internal/cache/cache.go (hunk 3 above)
Comment (mikehelmick) [reply internal/cache/cache.go:107]: for a read heavy cache - I think's it's best to use the RLock that can be shared.

File: internal/cache/cache_test.go (hunk 6 above)
Comment (mikehelmick) [reply internal/cache/cache_test.go:121]: how about 10?

**Pull request merged**